libcxx/include/__exception/exception_ptr.h

```diff
@@ -65,8 +65,11 @@ class _LIBCPP_EXPORTED_FROM_ABI exception_ptr {
   friend _LIBCPP_HIDE_FROM_ABI exception_ptr make_exception_ptr(_Ep) _NOEXCEPT;
 
 public:
-  // exception_ptr is basically a COW string.
+  // exception_ptr is basically a COW string so it is trivially relocatable.
+  // However, it's not replaceable because destroying and move-constructing could cause
+  // the underlying refcount to hit 0 if we're self-assigning.
```

Comment:
I'm not sure I understand this. If we're self-assigning it's a no-op and shouldn't potentially delete the underlying resource.

Reply:
The semantics of `__is_replaceable` mean that move-assignment is equivalent to destroy followed by a move construction. However, you can't replace a move-assignment by a destroy + move-construct for `std::exception_ptr`, since if both the source and destination are the same object, the destroy + move-construction will result in the refcount hitting 0 (after destruction) whereas the move-assignment wouldn't have.

This actually makes me realize: you can never do destroy + move-construct if the source and the destination are the same object. This has nothing to do with the presence of a refcount in the object. How is `__is_replaceable` supposed to work in that case? Aren't *all* types non-replaceable, then?

Reply:
I don't think you have to consider the same memory location case. You can't relocate an object into the same place it is already in, since you can't relocate from an object that is already destroyed (which you just did by relocating into the same memory location).

Reply:
That makes me wonder how useful `__is_replaceable` is as a trait outside of relocation use cases. It might not be.

In order to make progress for the time being, I will add documentation clarifying that `__is_replaceable` doesn't apply if you're doing a self-assignment. And then I will revisit whether these types can be replaceable or not based on that new point of view.